`ot_entropy_src`: unify implementations from EarlGrey and Darjeeling

rivos-eblot opened the pull request:
Support `ot_entropy_src` v2 and v3, using a property to select the emulated version.

Rework & clean up random/noise sources management, now that both tops are using the same entropy stack.

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: unify EarlGrey and Darjeeling versions

- remove deprecated Darjeeling version
- rename EarlGrey version as unique entropy src implementation
- update machine definitions and dependencies

Note that EarlGrey-1.0.0 version uses v2 while Darjeeling master uses v3,
so execution of Darjeeling machine is broken for now, till the implementation
also supports v3

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ast_eg: implement the noise provider interface

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_noise_src: define a new interface for noise provider

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: use the new noise provider interface

replace the direct ot_ast_eg calls with the new abstract interface

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: remove random interface abstraction.

Both EarlGrey and Darjeeling now use the EntropySrc to obtain entropy,
whereas previous Darjeeling implementation was collecting entropy from
the AST.

Rework the EntropySrc class to expose a method to provide entropy to the
CSRNG, and update CSRNG to use this "unique" entropy source.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_csrng: replace ot_random_src_if calls with direct ot_entropy_src calls

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ast_dj: remove ot_random_src interface implementation

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ast_dj: implement the noise provider interface

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_random_src: remove ot_random_src implementation

it is no longer useful

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_darjeeling: update entropy stack configuration

- configure entropy src version
- replace random src interface with entropy src
- replace entropy src noise source

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_earlgrey: update entropy stack configuration

- configure entropy src version
- replace random src interface with entropy src
- replace entropy src noise source

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_otp: remove useless entropy API

Move the en_csrng_sw_app_read field to where it belongs: OtOTPHWCfg

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_otp: make multibool bit type more visible

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_common: define type aliases for OpenTitan multibit booleans

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv, hw/opentitan: all: fix dash vs. underscore link names

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

AlexJones0 commented on hw/riscv/ot_darjeeling.c:
```diff
@@ -745,45 +745,23 @@ static const IbexDeviceDef ot_dj_soc_devices[] = {
         .gpio = IBEXGPIOCONNDEFS(
-            OT_DJ_SOC_GPIO_ALERT(0, 35),
-            OT_DJ_SOC_GPIO_ALERT(1, 36)
+         OT_DJ_SOC_GPIO_SYSBUS_IRQ(0, PLIC, 89),
+         OT_DJ_SOC_GPIO_SYSBUS_IRQ(1, PLIC, 90),
+         OT_DJ_SOC_GPIO_SYSBUS_IRQ(2, PLIC, 91),
+         OT_DJ_SOC_GPIO_SYSBUS_IRQ(3, PLIC, 92),
+         OT_DJ_SOC_GPIO_ALERT(0, 35),
+         OT_DJ_SOC_GPIO_ALERT(1, 36)
```
Are these correct? On Darjeeling I think `entropy_src` should be [PLIC IRQ IDs 121-124](https://github.com/lowRISC/opentitan/blob/6e53d12a9140357c203558ab3ee28269e5cbd4d4/hw/top_darjeeling/sw/autogen/top_darjeeling.h#L1112-L1115) and [alert IDs 64-65](https://github.com/lowRISC/opentitan/blob/6e53d12a9140357c203558ab3ee28269e5cbd4d4/hw/top_darjeeling/sw/autogen/top_darjeeling.h#L1306-L1307).

rivos-eblot replied:
Good catch, it seems "our" Darjeeling top is not aligned anymore with Darjeeling master here :cry:

rivos-eblot replied:
... or rather the whole map is not the proper one 😨

AlexJones0 replied:
I guessed as much given that the indexes were so different :sweat_smile:

rivos-eblot replied:
I think it is better to address this point in another PR then, as at least these values are coherent with "some" Darjeeling version, otherwise the IRQ lines might be used twice :(

rivos-eblot replied:
Created an issue to track this update: https://github.com/lowRISC/qemu/issues/265

AlexJones0 reviewed (commented):
(only looked at the first commit for now because I didn't realise it was still a draft)

rivos-eblot commented:
> (only looked at the first commit for now because I didn't realise it was still a draft)

I was eagerly waiting for the results from the EG CI before removing the flag :-)

rivos-eblot commented:
It seems there is one single test failing: 
```
//sw/device/silicon_creator/rom_ext/e2e/ownership:invalid_key_alg_activate_request_test_sim_qemu_rom_ext
```

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -0,0 +1,2006 @@
+REG32(REGWEN, 0x18u)
+    FIELD(REGWEN, EN, 0u, 1u)
+
+/* second group og registers, "rev", only used with v2 */
```
```suggestion
/* second group of registers, "rev", only used with v2 */
```

rivos-eblot commented:
(rebase after OT_OTP update)

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -0,0 +1,2006 @@
+static_assert(ALERT_COUNT == NUM_ALERTS, "Invalid alert count");
+
+typedef enum {
+    ENTROPY_SRC_EG_IDLE,
```
Might also be worth doing a final commit with `s/entropy_src_eg/entropy_src` now that the implementations are unified?

rivos-eblot replied:
I thought I did, I missed the enumerated values sorry about that.

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -410,6 +498,7 @@ struct OtEntropySrcClass {
     ResettablePhases parent_phases;
 };
 
+/* @todo: need to check the FSM code are identical on v2 and v3... */
```
With regards to this TODO: as far as I can tell both versions have identical FSM codes, but these do not match the current QEMU constants. Specifically `ENTROPY_SRC_EG_SHA3_PREP` does not exist in the RTL. This causes the subsequent `SHA3_PROCESS` to `ALERT_HANG` FSM state codes to differ. Also, the `ENTROPY_SRC_EG_ERROR` code differs from any current state, as its code is `0bb001110011`.

rivos-eblot replied:
I've updated the FSM and removed `SHA3_PREP` and `SHA3_QUIESCE`

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -1186,16 +1353,14 @@ ot_entropy_src_regs_read(void *opaque, hwaddr addr, unsigned size)
-            REG_MB4_IS_TRUE(s, CONF, ENTROPY_DATA_REG_ENABLE) &&
+            ((s->version < 3) ?
+                 REG_MB4_IS_TRUE(s, hi, CONF, V2_ENTROPY_DATA_REG_ENABLE) :
+                 REG_MB4_IS_TRUE(s, hi, CONF, V3_ENTROPY_DATA_REG_ENABLE)) &&
```
Nit: formatting is getting kind of messy here, maybe this `ENTROPY_DATA_REG` check should be extracted to its own function? And the same with the the common `REG_HI_V{2,3}_BASE` pattern?

rivos-eblot replied:
You are very right, it would be more readable.

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -164,13 +203,16 @@ REG32(DEBUG_STATUS, 0xd0u)
     FIELD(RECOV_ALERT_STS, MODULE_ENABLE_FIELD_ALERT, 2u, 1u)
     FIELD(RECOV_ALERT_STS, THRESHOLD_SCOPE_FIELD_ALERT, 3u, 1u)
     FIELD(RECOV_ALERT_STS, RNG_BIT_ENABLE_FIELD_ALERT, 5u, 1u)
+    /* bitfield only defined on version 2 */
```
Is this correct? For both [`earlgrey_1.0.0`](https://opentitan.org/earlgrey_1.0.0/book/hw/ip/entropy_src/doc/registers.html#recov_alert_sts) and [`master`](https://opentitan.org/book/hw/ip/entropy_src/doc/registers.html#recov_alert_sts) this field is `SHA3_START` and not `INSERT_START`. This may have just been a rename in the past?

rivos-eblot replied:
You are right, I don't know where I got this value from, maybe a very old EG version. Fixed.
Ok I got it, me bad; see next conversation for details.

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -1417,9 +1666,10 @@ static void ot_entropy_src_regs_write(void *opaque, hwaddr addr, uint64_t val64,
-        CHECK_MULTIBOOT(s, FW_OV_SHA3_START, FW_OV_INSERT_START);
-        if (REG_MB4_IS_TRUE(s, FW_OV_SHA3_START, FW_OV_INSERT_START)) {
+        s->regs_hi[reg] = val32;
+        // @todo need to ckeck v2 vs. v3
```
QEMU comment style (`/* ... */`) and `ckeck` -> `check`. Aside: what specifically needs to be checked here - just the RTL behaviour?

rivos-eblot replied:
Actually nothing. I got caught myself with my own hack *and* the laziness of not adding a comment, then forgetting about the whole thing:

Entropy Src alert bits are so many that I implemented a macro, `CHECK_MULTIBOOT` that:
1. Compare the MB4 of a specified register to a MB4 value
2. If the MBs differ, it emits a warning, then flag the error in the R_RECOV_ALERT register
This macro saves a lot of boring boilerplate code that deals with unbelievly long and (I think) confusing bitfields such as 
`R_RECOV_ALERT_STS_ENTROPY_DATA_REG_ENABLE_FIELD_ALERT_MASK`

This macro is used many times, and work well ... till a single exception: there is one register whose matching error bitfield differ in naming: `FW_OV_SHA3_START. FW_OV_INSERT_START` trigger error `RECOV_ALERT_STS.FW_OV_SHA3_START_FIELD_ALERT` not `RECOV_ALERT_STS. FW_OV_INSERT_START_FIELD_ALERT`

Rather that coming up with a fully different implementation for this very specific exception, I've renamed the bit in the `RECOV_ALERT_STS` register, so that the `CHECK_MULTIBOOT` could work without trouble.

However, I deeply failed to document this workaround, then re-reading the code back one year or so later with all the EG 0.2.5.2, EG 1.0, EG-master, DJ-*, I got caugth with my own workaround and interpreted it as a change between EG and DJ version. 😬

I've updated the code as follows:
1. define an alias (#define) for the register bitfield rather than renaming it
2. added a comment why this alias exist
3. removed the v2/v3 related todo

AlexJones0 commented on include/hw/opentitan/ot_noise_src.h:
```diff
@@ -26,12 +26,36 @@
+    /*
+     * Fill up a buffer with noise data
+     *
+     * @buf destination buffer
```
```suggestion
     * @return the fill rate, in bytes per second
     */
    unsigned (*get_fill_rate)(OtNoiseSrcIf *dev);

    /*
     * Fill up a buffer with noise data
     *
     * @buffer destination buffer
```

AlexJones0 commented on hw/opentitan/ot_ast_eg.c:
```diff
@@ -260,6 +253,21 @@ static void ot_ast_eg_clock_ext_freq_select(OtClockCtrlIf *dev, bool enable)
+{
+    (void)dev;
+
+    return OT_AST_EG_NOISE_4BIT_RATE * 2u; /* 4 bits to byte */
```
Not quite following this, but maybe I'm too tired. Is this implying that because we get 4 bits of noise at a rate of 50 kHz, we will get 1 byte of noise at a rate of 100 kHz? Should it not be `// 2`?

rivos-eblot replied:
No it is just me chosing to use rate in bits, then changing for bytes, and messing everything up :-) Good catch.
Note that this value comes from the HW description, and we might need to adjust it with some real guest measurement.

AlexJones0 commented on hw/opentitan/ot_ast_dj.c:
```diff
@@ -133,6 +134,8 @@ static const char REGB_NAMES[REGSB_COUNT][6U] = {
 };
 #undef REG_NAME_ENTRY
 
+#define OT_AST_DJ_NOISE_BITRATE 1000000u /* 1 MHz, need to check with HW */
```
Maybe add a `todo` so this isn't lost?

rivos-eblot replied:
Sure. Also adding it as a property, since there may be several "Darjeeling" versions.

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: update FSM

Remove deprecated states, update state magic values

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

AlexJones0 commented on hw/opentitan/ot_entropy_src.c:
```diff
@@ -0,0 +1,2000 @@
+                          TYPE_OT_ENTROPY_SRC "-regs-hi", REGS_HI_SIZE);
+    memory_region_add_subregion(&s->mmio,
+                                s->version < 3 ? REGS_HI_V2_BASE :
+                                                 REGS_HI_V3_BASE,
```
Might it also be worth making this common `REG_HI_V{2,3}_BASE` pattern its own function, as this is repeated a lot and can complicate the formatting in many places?

AlexJones0 reviewed (approved):
Thanks, this LGTM. The CI failure in `newversion_pq_downgrade_test` looks unrelated (albeit I'm not sure why it failed, it doesn't look like the normal timing error) as it passes fine when I test locally.

rivos-eblot committed:
[ot] .gitlab-ci.d/opentitan: ot-smoke.yml: fix typo for Darjeeling smoke test

Introduced in 5d9dfd633

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: add support for multiple versions

Support either version 2 or 3. Both versions are mostly similar, however a "REV" register
has been added in v2 then removed in v3. This register stands at the 8th register slot.
This causes most of the register map to shift down on v3 compared to v2. The version
upgrade also introduces a couple of bitfield permutations and width changes in other
registers.

Add a new property to select the version to use.

To address the register map shifting, split the register map in 3 segments:
* lo: for registers before the REV register, whose base address is the same for both
  versions
* rev: for the register which has been introduced in v2 and removed in v3
* hi: for the registers located after the REV register, whose base address depends
  on the presence of the REV register, i.e. whether emulutating v2 or v3
Use QEMU sub memory regions to map each register segment to the proper location based
on the selected version. The addresses of each segment defined with REG32 macros are
made relative to the start location of the matching segment. Traces reporting addresses
are updated with the base address of each segment to preserve the "natural" address of
each register. In order to avoid many offset computations and degrade code readability,
also split the register array to support on array for each segment.

Update macros to use the select the proper segment while addressing the register content.

To address the register bitfield permutations/widening, as the scope of these changes is
limited, simply dispatch handling based on the selected version.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>
